<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# **Hands-on Lab: Interactive Visual Analytics with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [39]:
# Local Jupyter setup (no piplite/js required for standard Jupyter environments)
import folium
import pandas as pd
from folium.plugins import MarkerCluster, MousePosition
from folium.features import DivIcon

# Uncomment below if you need to install packages:
# !pip install folium pandas

In [40]:
import folium
import pandas as pd

In [41]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/v4/DV0101EN-Exercise-Generating-Maps-in-Python.ipynb)


## Task1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [42]:
# Download and read the `spacex_launch_geo.csv`
URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
spacex_df = pd.read_csv(URL)
spacex_df.head()

,Flight Number,Date,Time (UTC),Booster Version,Launch Site,Payload,Payload Mass (kg),Orbit,Customer,Landing Outcome,class,Lat,Long
0,1,2010-06-04,18:45:00,F9 v1.0 B0003,CCAFS LC-40,Dragon Spacecraft Qualification Unit,0.0,LEO,SpaceX,Failure (parachute),0,28.562302,-80.577356
1,2,2010-12-08,15:43:00,F9 v1.0 B0004,CCAFS LC-40,"Dragon demo flight C1, two CubeSats, barrel o...",0.0,LEO (ISS),NASA (COTS) NRO,Failure (parachute),0,28.562302,-80.577356
2,3,2012-05-22,7:44:00,F9 v1.0 B0005,CCAFS LC-40,Dragon demo flight C2+,525.0,LEO (ISS),NASA (COTS),No attempt,0,28.562302,-80.577356
3,4,2012-10-08,0:35:00,F9 v1.0 B0006,CCAFS LC-40,SpaceX CRS-1,500.0,LEO (ISS),NASA (CRS),No attempt,0,28.562302,-80.577356
4,5,2013-03-01,15:10:00,F9 v1.0 B0007,CCAFS LC-40,SpaceX CRS-2,677.0,LEO (ISS),NASA (CRS),No attempt,0,28.562302,-80.577356


Now, you can take a look at what are the coordinates for each site.


In [43]:
# Get unique launch sites with coordinates (no 'class' column in this dataset)
launch_sites_df = spacex_df[['Launch Site', 'Lat', 'Long']].drop_duplicates()
launch_sites_df

,Launch Site,Lat,Long
0,CCAFS LC-40,28.562302,-80.577356
26,VAFB SLC-4E,34.632834,-120.610745
36,KSC LC-39A,28.573255,-80.646895
49,CCAFS SLC-40,28.563197,-80.576820


Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [ ]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [ ]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)
site_map

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [ ]:
# Initialize map centered on NASA JSC
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)

# Add circles and labels for each launch site
for lat, lng, label in zip(launch_sites_df['Lat'], launch_sites_df['Long'], launch_sites_df['Launch Site']):
    # Circle marker
    folium.Circle(
        [lat, lng],
        radius=1000,
        color='#d35400',
        fill=True,
        fill_color='#d35400',
        fill_opacity=0.7,
        popup=label
    ).add_to(site_map)
    
    # Text label
    folium.map.Marker(
        [lat, lng],
        icon=DivIcon(
            icon_size=(20,20),
            icon_anchor=(0,0),
            html=f'<div style="font-size:12; color:#d35400;"><b>{label}</b></div>'
        )
    ).add_to(site_map)

# Display the map
site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png">
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


In [ ]:
# Task 2: Mark success (green) and failed (red) launches for each site
# Load full launch dataset from earlier EDA
launch_data_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/dataset_part_2.csv"
launch_df = pd.read_csv(launch_data_url)

# Fix column name mismatch (launch_df uses 'LaunchSite' without space, geo df uses 'Launch Site')
launch_df = launch_df.rename(columns={'LaunchSite': 'Launch Site'})

# Normalize site name formatting (replace hyphens with spaces to match geo dataset)
launch_df['Launch Site'] = launch_df['Launch Site'].str.replace('-', ' ', regex=False)
spacex_df['Launch Site'] = spacex_df['Launch Site'].str.replace('-', ' ', regex=False)

# Merge to get coordinates for each launch
launch_geo_df = pd.merge(
    launch_df,
    spacex_df[['Launch Site', 'Lat', 'Long']],
    on='Launch Site',
    how='left'
).dropna(subset=['Lat', 'Long'])

# Create marker cluster for launches
marker_cluster = MarkerCluster().add_to(site_map)

# Add success (green) and failed (red) markers
for _, row in launch_geo_df.iterrows():
    color = 'green' if row['Class'] == 1 else 'red'
    folium.CircleMarker(
        location=[row['Lat'], row['Long']],
        radius=5,
        color=color,
        fill=True,
        fill_color=color,
        fill_opacity=0.7,
        popup=f"Site: {row['Launch Site']}<br>Outcome: {'Success' if row['Class']==1 else 'Failure'}"
    ).add_to(marker_cluster)

# Add legend
legend_html = '''
<div style="position:fixed; bottom:50px; left:50px; width:150px; height:90px;
background-color:white; border:2px solid grey; z-index:9999; font-size:14px; padding:10px;">
<b>Launch Outcome</b><br>
<span style="color:green;">●</span> Success<br>
<span style="color:red;">●</span> Failure
</div>
'''
site_map.get_root().html.add_child(folium.Element(legend_html))
site_map

In [ ]:
# Task 3: Calculate distances between launch sites and mark them on the map
from math import radians, sin, cos, sqrt, atan2

def haversine(lat1, lon1, lat2, lon2):
    """Calculate distance between two coordinates in kilometers"""
    R = 6371  # Earth radius in km
    lat1, lon1, lat2, lon2 = map(radians, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c = 2 * atan2(sqrt(a), sqrt(1-a))
    return R * c

# Calculate pairwise distances between unique launch sites
print("Pairwise distances between launch sites (km):")
sites = launch_sites_df[['Launch Site', 'Lat', 'Long']].values.tolist()
for i in range(len(sites)):
    for j in range(i+1, len(sites)):
        site1, lat1, lon1 = sites[i]
        site2, lat2, lon2 = sites[j]
        dist = haversine(lat1, lon1, lat2, lon2)
        print(f"{site1} ↔ {site2}: {dist:.1f} km")

# Add distance lines between sites on the map
for i in range(len(sites)):
    for j in range(i+1, len(sites)):
        site1, lat1, lon1 = sites[i]
        site2, lat2, lon2 = sites[j]
        folium.PolyLine(
            locations=[[lat1, lon1], [lat2, lon2]],
            color='blue',
            weight=2,
            opacity=0.5,
            popup=f"Distance: {haversine(lat1, lon1, lat2, lon2):.1f} km"
        ).add_to(site_map)

site_map

### Analysis Findings
1. **Proximity to Equator**: 
   - CCAFS LC-40, CCAFS SLC-40, and KSC LC-39A are at ~28.5°N (low latitude, favorable for launches due to higher rotational speed)
   - VAFB SLC-4E is at ~34.6°N (higher latitude, not near the equator)
   - Most sites are at relatively low latitudes compared to higher northern latitudes, optimizing launch efficiency.

2. **Proximity to Coast**:
   - All 4 launch sites are directly located on the coast (Florida sites on the Atlantic coast, VAFB on the Pacific coast)
   - This is critical for safety (debris falls into water) and for ocean landings of Falcon 9 first stages.

3. **Distances Between Sites**:
   - Florida-based sites (CCAFS LC-40, CCAFS SLC-40, KSC LC-39A) are all within ~7 km of each other
   - VAFB SLC-4E is ~3,700 km from the Florida launch sites